### PR TITLE
feat: gateway swap between federations

### PR DESCRIPTION
**Pull request by okjodom**
When a gateway is connected to any two federations, it can facilitate a direct e-cash swap between these federations, without routing over lightning.

## Specifics:

`GatewayPayInvoice` state machine checks if a proposed payment can be settled via direct swap. Specifically an LN payment that can be swapped will _have the gateway node pubkey_ and _a short channel id assigned by this gateway_ to one of it's connected federations in the route hint last hop. If these markers are present, the gateway instructs the client for the receiving federation to buy the preimage, and in turn, settles the payment in the originating federation client

### CODE REVIEW

File: fedimint-client/src/lib.rs [hunk 1]
```diff
@@ -455,6 +455,7 @@ fn states_add_instance(
     })
 }
 
+#[derive(Debug)]
```
Comment (m1sterc001guy) [inline fedimint-client/src/lib.rs:458]: Is this necessary in this PR? Seems unrelated

File: gateway/ln-gateway/src/state_machine/mod.rs [hunk 2]
```diff
@@ -132,7 +133,13 @@ pub trait GatewayClientExt {
     /// Attempt fulfill HTLC by buying preimage from the federation
     async fn gateway_handle_intercepted_htlc(&self, htlc: Htlc) -> anyhow::Result<OperationId>;
 
-    /// Subscribe to updates when the gateway is handling an intercepted HTLC
+    async fn gateway_handle_direct_swap(
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/state_machine/mod.rs:136]: Nit: add comment on this function too

File: gateway/ln-gateway/src/state_machine/mod.rs [hunk 3]
```diff
@@ -627,3 +693,22 @@ impl TryFrom<InterceptHtlcRequest> for Htlc {
...
+    payment_hash: sha256::Hash,
+    amount_msat: Amount,
+}
+
+impl TryFrom<lightning_invoice::Bolt11Invoice> for SwapParameters {
+    type Error = anyhow::Error;
+
+    fn try_from(s: lightning_invoice::Bolt11Invoice) -> Result<Self, Self::Error> {
+        let payment_hash = *s.payment_hash();
+        let amount_msat = s.amount_milli_satoshis().map(Amount::from_msats).unwrap();
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/state_machine/mod.rs:708]: Return this `Err` instead of unwrap

File: gateway/ln-gateway/src/state_machine/pay.rs [hunk 4]
```diff
@@ -136,6 +140,11 @@ pub enum OutgoingPaymentError {
         error: OutgoingContractError,
         contract: OutgoingContractAccount,
     },
+    #[error("An error occurred while attempting direct swap.")]
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/state_machine/pay.rs:143]: Nit: "direct swap between federations"

File: gateway/ln-gateway/src/state_machine/pay.rs [hunk 5]
```diff
@@ -347,6 +435,26 @@ impl GatewayPayInvoice {
             invoice,
         })
     }
+
+    async fn check_swap_to_federation(
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/state_machine/pay.rs:439]: Nit: lets add some comments here

File: gateway/ln-gateway/src/state_machine/pay.rs [hunk 6]
```diff
@@ -242,20 +251,89 @@ impl GatewayPayInvoice {
...
+        let mut preimage_sub = client
+            .gateway_subscribe_ln_receive(operation_id)
+            .await
+            .map_err(|e| OutgoingPaymentError::SwapFailed {
+                contract: contract.clone(),
+                swap_error: format!("Failed to subscribe to direct swap: {}", e),
+            })?
+            .into_stream();
+
+        while let Some(state) = preimage_sub.next().await {
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/state_machine/pay.rs:281]: I don't think we want to wait on the output of the receive state machine in the transition function. If we detect that this should be an internal swap, I think we should spawn the state machine, but transition to a new state that explicitly waits on the state of the child sm in the trigger function. We do this in `complete.rs` as well.

We don't want to block in the transition function ideally since we're holding a lock and no other state machines can make progress. The only reason we do `pay` for external lightning payments in the transition function is because `pay` is not idempotent.

**Review by m1sterc001guy — CHANGES_REQUESTED** [review]
Mostly nits, but one issue regarding when we do the direct swap IMO

File: fedimint-client/src/lib.rs (hunk 1 above)
Comment (okjodom) [reply fedimint-client/src/lib.rs:458]: necessary because we reference `FederationToClientMap` within `GatewayClientGen` which is debug

File: gateway/ln-gateway/tests/integration_tests.rs [hunk 7]
```diff
@@ -962,6 +962,77 @@ async fn test_gateway_shows_balance_for_any_connected_federation() -> anyhow::Re
...
+                    assert_matches!(waiting_funds, LnReceiveState::AwaitingFunds { .. });
+                    let claimed = receive_sub.ok().await?;
+                    assert_matches!(claimed, LnReceiveState::Claimed);
+                    assert_eq!(client2.get_balance().await, msats(2_500));
+                }
+                _ => panic!("Expected Lightning payment!"),
+            }
+
+            // Check gateway balances after facilitating direct swap between federations
+            let post_balances = get_balances(&rpc, &[id1, id2]).await;
```
[the remark this answers is not in the document]
Comment (okjodom) [reply gateway/ln-gateway/tests/integration_tests.rs:1026]: added check below

File: gateway/ln-gateway/src/state_machine/pay.rs [hunk 8]
```diff
@@ -242,21 +264,75 @@ impl GatewayPayInvoice {
...
+                                        contract,
+                                        federation_id,
+                                        operation_id,
+                                    },
+                                )),
+                            };
+                        }
+                        Err(e) => {
+                            tracing::debug!("Failed to buy preimage via direct swap: {}", e)
+                            // Fall back to lightning payments
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/state_machine/pay.rs:324]: Hmm do we want to fall back here? If the node in the route hint matches the gateway node, that means that the lightning node is always going to try and do a self payment, which we know won't work. Maybe we should just fail?

File: gateway/ln-gateway/src/state_machine/pay.rs [hunk 9]
```diff
@@ -409,6 +528,105 @@ impl GatewayPayClaimOutgoingContract {
...
+pub struct GatewayPayWaitForSwapPreimage {
+    contract: OutgoingContractAccount,
+    federation_id: FederationId,
+    operation_id: OperationId,
+}
+
+impl GatewayPayWaitForSwapPreimage {
+    fn transitions(
+        &self,
+        _global_context: DynGlobalClientContext,
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/state_machine/pay.rs:541]: You can just remove this parameter

File: gateway/ln-gateway/src/state_machine/pay.rs (hunk 8 above)
Comment (okjodom) [reply gateway/ln-gateway/src/state_machine/pay.rs:324]: I was undecided on the desired behavior. I agree failing here makes more sense

**Review by m1sterc001guy — APPROVED** [review]

**Review by m1sterc001guy — APPROVED** [review]
